brave/src/test/java/brave/features/propagation/SecondarySamplingTest.java

```diff
@@ -13,195 +13,353 @@
+ * TraceContext#sampled()} at the beginning of the trace, and propagated downstream in B3 headers.
+ * This simplifies the question of what to do about Edge who make a constant decision, and Triage,
+ * who make a conditional decision anywhere. Since they are not the primary deciders of the trace,
+ * they cannot affect {@link TraceContext#sampled()}.
+ *
+ * <p>This implies Edge and Triage use secondary, extra fields, to propagate state about their
+ * respective decisions. When Edge or Triage decides to record data, they must set {@link
+ * TraceContext#sampledLocal()} so that recording occurs even if the primary decision is unsampled.
+ * At startup, {@link Tracing.Builder#alwaysReportSpans()} ensures the trace forwarder gets data
+ * regardless of the B3 decision.
+ *
+ * <p>Secondary fields and state management about them is the responsibility of {@link
+ * ExtraFieldPropagation}. Here, you can add an extra field that corresponds to a header sent out of
+ * process. You can also add an {@link ExtraFieldPlugin} to ensure a relevant tag gets to the trace
+ * forwarder.
+ *
+ * TODO: continue to write
+ */
 public class SecondarySamplingTest {
-  List<zipkin2.Span> zipkin = new ArrayList<>();
-  List<MutableSpan> edge = new ArrayList<>(), links = new ArrayList<>(), triage =
-    new ArrayList<>();
-  FinishedSpanHandler triageHandler = new FinishedSpanHandler() {
-    @Override public boolean handle(TraceContext context, MutableSpan span) {
-      return triage.add(span);
-    }
+  Callback<Void> noopCallback = mock(Callback.class);
+  InMemoryStorage zipkin = InMemoryStorage.newBuilder().build();
+  InMemoryStorage edge = InMemoryStorage.newBuilder().build();
+  InMemoryStorage triage = InMemoryStorage.newBuilder().build();
+
+  // pretend this is a dynamic configuration store
+  SetMultimap<String, String> triageKeyToService = new SetMultimap<>();
+
+  Function<String, ExtraFieldCustomizer> configureEdgeSampling = localServiceName -> builder -> {
+    builder.addPlugin(new EdgeSamplingPlugin(localServiceName));
+    builder.addPlugin(SamplingSystemsPlugin.get());
+  };
+  Function<String, ExtraFieldCustomizer> configureTriageSampling = localServiceName -> builder -> {
+    builder.addPlugin(new TriageSamplingPlugin(triageKeyToService, localServiceName));
+    builder.addPlugin(SamplingSystemsPlugin.get());
+  };
+  Function<String, ExtraFieldCustomizer> configureAllSampling = localServiceName -> builder -> {
+    configureEdgeSampling.apply(localServiceName).customize(builder);
+    configureTriageSampling.apply(localServiceName).customize(builder);
   };
 
-  SecondarySampling secondarySampling = SecondarySampling.create()
-    .putSystem("edge", new FinishedSpanHandler() {
-      @Override public boolean handle(TraceContext context, MutableSpan span) {
-        return edge.add(span);
-      }
-    })
-    .putSystem("links", new FinishedSpanHandler() {
+  Reporter<zipkin2.Span> traceForwarder = span -> {
```

Comment:
@jeqo this is what I was thinking it would be nice to have a trace forwarder be able to do. tag-based routing to push one span to potentially multiple backends

Reply:
love it. it's kind of what the kafka-storage tries to achieve by exposing spans as a topic; but for every back-end. Would require changes on the server side tho, reminds me the idea of filters: should we revive https://github.com/openzipkin/zipkin/pull/2072 ? :)

Reply:
actually, instead of on the server side, we can do it only on the forwarder span consumer, using the same filter principle.

Reply:
yeah :) that's what I was thinking, too. there's no need to couple this to a server change yet. however, it is probably still a good idea to pick up the trace scoped types discussed earlier, albeit in a different repo.